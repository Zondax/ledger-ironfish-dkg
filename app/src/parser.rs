mod asset_identifier;
mod constants;
mod error;
mod from_bytes;
mod memo;
mod merkle_note;
mod note;
mod object_list;
mod transaction;
mod tx_version;

pub use asset_identifier::AssetIdentifier;
pub use constants::*;
pub use error::ParserError;
pub use from_bytes::FromBytes;
pub use memo::Memo;
pub use merkle_note::MerkleNote;
pub use note::Note;
pub use object_list::ObjectList;
pub use transaction::{Burn, Mint, Output, Spend, Transaction};
pub use tx_version::TransactionVersion;
