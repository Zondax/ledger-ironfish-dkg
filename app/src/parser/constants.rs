
// Value commitment
// root of the tree
// tree_size
// nullifier
// Authorizing signature
//

pub const PUBLIC_ADDRESS_SIZE: usize = 32;
// 32-bytes public_key randomness
// 192-byes proof
// 32-bytes value_commitment
// 32-bytes root hash
// 4-bytes tree_size
// 32-bytes nullifier
// 64-bytes authorize signatures
pub const SPEND_LEN: usize = 32 + 192 + 32 + 32 + 4 + 32 + 64;
// 192-bytes proof + 328-bytes Merkle Note
pub const OUTPUT_LEN: usize = 192 + 328;
// 33-bytes public_key randomness
// 192-byes proof
// 193-bytes asset_len
// 8-bytes value
// + optional values 32 bytes owner, opion flag + 32_bytes new owner
pub const MINT_LEN: usize = 32 + 192 + ASSET_LEN + 8;
pub const BURN_LEN: usize = 32 + 8;
// Asset len description
// 32-bytes creator(address)
// 32-bytes(name)
// 96-bytes(metadata_len)
// 1-byte(nonce)
pub const ASSET_LEN: usize = 161; //193;
pub const REDJUBJUB_SIGNATURE_LEN: usize = 64;
pub const KEY_LENGTH: usize = 32;
pub const SCALAR_SIZE: usize = 32;
pub const MEMO_SIZE: usize = 32;
pub const AMOUNT_VALUE_SIZE: usize = 8;

pub const PLAINTEXT_NOTE_SIZE: usize = PUBLIC_ADDRESS_SIZE
    + ASSET_ID_LENGTH
    + AMOUNT_VALUE_SIZE
    + SCALAR_SIZE
    + MEMO_SIZE
    + PUBLIC_ADDRESS_SIZE;

/// Length in bytes of the asset identifier
pub const ASSET_ID_LENGTH: usize = 32;

pub const MAC_SIZE: usize = 16;
// Size of a merkle note
// https://github.com/iron-fish/ironfish/blob/master/ironfish-rust/src/note.rs#L30
pub const ENCRYPTED_NOTE_SIZE: usize =
    SCALAR_SIZE + MEMO_SIZE + AMOUNT_VALUE_SIZE + ASSET_ID_LENGTH + PUBLIC_ADDRESS_SIZE;

pub const ENCRYPTED_SHARED_KEY_SIZE: usize = 64;

pub const NOTE_ENCRYPTION_KEY_SIZE: usize = ENCRYPTED_SHARED_KEY_SIZE + MAC_SIZE;

pub const AFFINE_POINT_SIZE: usize = 32;

/// BLAKE2s personalization for deriving asset identifier from asset name
pub const ASSET_ID_PERSONALIZATION: &[u8; 8] = b"ironf_A_";

/// BLAKE2s personalization for PRF^nf = BLAKE2s(nk | rho)
pub const PRF_NF_PERSONALIZATION: &[u8; 8] = b"ironf_nf";

/// BLAKE2s personalization for the value commitment generator for the value
pub const VALUE_COMMITMENT_GENERATOR_PERSONALIZATION: &[u8; 8] = b"ironf_cv";

pub const TX_HASH_LEN: usize = 32;
pub const TRANSACTION_SIGNATURE_VERSION: &[u8; 1] = &[0];
pub const SIGNATURE_HASH_PERSONALIZATION: &[u8; 8] = b"IFsighsh";
