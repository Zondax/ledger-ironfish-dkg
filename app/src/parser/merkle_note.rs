use core::ptr::addr_of_mut;
use group::{
    cofactor::{CofactorCurve, CofactorCurveAffine, CofactorGroup},
    prime::PrimeGroup,
    Curve, Group,
};

use crate::{
    crypto::{calculate_key_for_encryption_keys, parse_affine_point, read_fr, read_scalar},
    ironfish::{
        errors::IronfishError,
        view_keys::{shared_secret, OutgoingViewKey},
    },
    FromBytes,
};
use crate::{
    crypto::{decrypt, parse_extended_point},
    ironfish::public_address::PublicAddress,
    parser::{
        ParserError, AFFINE_POINT_SIZE, ENCRYPTED_NOTE_SIZE, MAC_SIZE, NOTE_ENCRYPTION_KEY_SIZE,
    },
};
use arrayref::array_ref;
use jubjub::AffinePoint;
use jubjub::{ExtendedPoint, Scalar, SubgroupPoint};
use nom::bytes::complete::take;

use super::{Note, ENCRYPTED_SHARED_KEY_SIZE};

#[derive(Clone, Debug)]
pub struct MerkleNote<'a> {
    /// Randomized value commitment. Sometimes referred to as
    /// `cv` in the literature. It's calculated by multiplying a value by a
    /// random number. Commits this note to the value it contains
    /// without revealing what that value is.
    /// Use AffinePoint instead of ExtendedPoint
    /// for simplicity and easy conversion from/to bytes
    pub(crate) value_commitment: AffinePoint,

    /// The hash of the note, committing to it's internal state
    pub(crate) note_commitment: Scalar,

    /// Public part of ephemeral diffie-hellman key-pair. See the discussion on
    /// [`shared_secret`] to understand how this is used
    // pub(crate) ephemeral_public_key: SubgroupPoint,
    // We use AffinePoint because it is more compact and memory efficient
    // we do not have the right API to parse bytes into a SubgroupPoint
    pub(crate) ephemeral_public_key: AffinePoint,

    /// note as encrypted by the diffie hellman public key
    /// we use this data to decrypt a Note as represented by:
    /// https://github.com/iron-fish/ironfish/blob/master/ironfish-rust/src/note.rs#L88
    pub(crate) encrypted_note: &'a [u8; ENCRYPTED_NOTE_SIZE + MAC_SIZE],

    /// Keys used to encrypt the note. These are stored in encrypted format
    /// using the spender's outgoing viewing key, and allow the spender to
    /// decrypt it. The receiver (owner) doesn't need these, as they can decrypt
    /// the note directly using their incoming viewing key.
    pub(crate) note_encryption_keys: &'a [u8; NOTE_ENCRYPTION_KEY_SIZE],
}
impl<'a> FromBytes<'a> for MerkleNote<'a> {
    #[inline(never)]
    fn from_bytes_into(
        input: &'a [u8],
        out: &mut core::mem::MaybeUninit<Self>,
    ) -> Result<&'a [u8], nom::Err<ParserError>> {
        let (rem, affine) = take(AFFINE_POINT_SIZE)(input)?;
        let affine = affine
            .try_into()
            .map_err(|_| ParserError::ValueOutOfRange)?;
        let value_commitment = parse_affine_point(affine)?;

        let (rem, raw_scalar) = take(3232usize)(rem)?;
        let raw_scalar = array_ref!(raw_scalar, 0, 32);

        let note_commitment = Scalar::from_bytes(raw_scalar)
            .into_option()
            .ok_or(ParserError::UnexpectedValue)?;

        // parsing ephemeral pubkey which is a SubgroupPoint
        let (rem, raw_scalar) = take(32usize)(rem)?;
        let raw_scalar = array_ref!(raw_scalar, 0, 32);

        // ephemeral_public_key is a subgroupPoint
        // however, we are read it as an extended point due to lack of support
        // to compute it from bytes, ironfish uses a custom version of the jubjub
        // crate that is incompatible with our target.
        let ephemeral_public_key = parse_affine_point(raw_scalar)?;

        // encrypted_note
        let (rem, raw_note) = take(ENCRYPTED_NOTE_SIZE + MAC_SIZE)(rem)?;
        let encrypted_note = array_ref!(raw_note, 0, ENCRYPTED_NOTE_SIZE + MAC_SIZE);

        // note_encryption_keys
        let (rem, encryption_keys) = take(NOTE_ENCRYPTION_KEY_SIZE)(rem)?;
        let note_encryption_keys = array_ref!(encryption_keys, 0, NOTE_ENCRYPTION_KEY_SIZE);

        let out = out.as_mut_ptr();

        unsafe {
            addr_of_mut!((*out).value_commitment).write(value_commitment);
            addr_of_mut!((*out).note_commitment).write(note_commitment);
            addr_of_mut!((*out).ephemeral_public_key).write(ephemeral_public_key);
            addr_of_mut!((*out).encrypted_note).write(encrypted_note);
            addr_of_mut!((*out).note_encryption_keys).write(note_encryption_keys);
        }
        Ok(rem)
    }
}

impl<'a> MerkleNote<'a> {
    pub fn decrypt_note_for_spender(
        &self,
        spender_key: &OutgoingViewKey,
    ) -> Result<Note, IronfishError> {
        let encryption_key = calculate_key_for_encryption_keys(
            spender_key,
            &self.value_commitment,
            &self.note_commitment,
            &self.ephemeral_public_key.to_bytes(),
        );

        let note_encryption_keys: [u8; ENCRYPTED_SHARED_KEY_SIZE] =
            decrypt(&encryption_key, self.note_encryption_keys)?;

        let public_address = PublicAddress::new(&note_encryption_keys[..32].try_into().unwrap())?;

        let (rem, secret_key) =
            read_fr(&note_encryption_keys[32..]).map_err(|_| IronfishError::InvalidScalar)?;
        let shared_key = shared_secret(&secret_key, &public_address.0, &self.ephemeral_public_key);
        let note =
            Note::from_spender_encrypted(public_address.0, &shared_key, &self.encrypted_note)?;

        // FIXME: Verify the node commitment
        // note.verify_commitment(self.note_commitment)?;

        Ok(note)
    }
}
