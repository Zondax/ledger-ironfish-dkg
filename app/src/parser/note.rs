use core::{mem::MaybeUninit, ptr::addr_of_mut};

use jubjub::AffinePoint;
use nom::number::complete::le_u64;

use crate::{
    crypto::{decrypt, read_scalar},
    ironfish::{errors::IronfishError, public_address::PublicAddress},
    parser::AssetIdentifier,
    FromBytes,
};

use super::{memo::Memo, ParserError, ENCRYPTED_NOTE_SIZE, MAC_SIZE};

/// A note (think bank note) represents a value in the owner's "account".
/// When spending, proof that the note exists in the tree needs to be provided,
/// along with a nullifier key that is made public so the owner cannot attempt
/// to spend that note again.447903
///
/// When receiving funds, a new note needs to be created for the new owner
/// to hold those funds.
pub struct Note {
    /// Asset identifier the note is associated with
    pub(crate) asset_id: AssetIdentifier,

    /// A public address for the owner of the note.
    pub(crate) owner: PublicAddress,

    /// Value this note represents.
    pub(crate) value: u64,

    /// A random value generated when the note is constructed.
    /// This helps create zero knowledge around the note,
    /// allowing the owner to prove they have the note without revealing
    /// anything else about it.
    pub(crate) randomness: jubjub::Fr,

    /// Arbitrary note the spender can supply when constructing a spend so the
    /// receiver has some record from whence it came.
    /// Note: While this is encrypted with the output, it is not encoded into
    /// the proof in any way.
    pub(crate) memo: Memo,

    /// A public address for the sender of the note.
    pub(crate) sender: PublicAddress,
}

impl Note {
    /// Create a note from its encrypted representation, given the spender's
    /// view key.
    ///
    /// The note is stored on the [`crate::outputs::OutputDescription`] in
    /// encrypted form. The spender encrypts it when they construct the output
    /// using a shared secret derived from the owner's public key.
    ///
    /// This function allows the owner to decrypt the note using the derived
    /// shared secret and their own view key.
    #[inline(never)]
    pub(crate) fn from_spender_encrypted(
        // public_address: SubgroupPoint,
        public_address: AffinePoint,
        shared_secret: &[u8; 32],
        encrypted_bytes: &[u8; ENCRYPTED_NOTE_SIZE + MAC_SIZE],
    ) -> Result<Self, IronfishError> {
        let mut this = MaybeUninit::uninit();
        Note::decrypt_note_parts(shared_secret, encrypted_bytes, &mut this)
            .map_err(|_| IronfishError::FailedXChaCha20Poly1305Decryption)?;

        let owner = PublicAddress(public_address);

        let out = this.as_mut_ptr();
        unsafe {
            addr_of_mut!((*out).owner).write(owner);
        }

        Ok(unsafe { this.assume_init() })
    }

    #[inline(never)]
    fn decrypt_note_parts(
        shared_secret: &[u8; 32],
        encrypted_bytes: &[u8; ENCRYPTED_NOTE_SIZE + MAC_SIZE],
        out: &mut MaybeUninit<Self>,
    ) -> Result<(), ParserError> {
        let out = out.as_mut_ptr();

        let plaintext_bytes: [u8; ENCRYPTED_NOTE_SIZE] =
            decrypt(shared_secret, encrypted_bytes).map_err(|_| ParserError::UnexpectedError)?;

        // Fr
        let (rem, randomness) = read_scalar(&plaintext_bytes[..])?;

        let (rem, value) = le_u64(rem)?;

        // Memo
        let memo = unsafe { &mut *addr_of_mut!((*out).memo).cast() };
        let rem = Memo::from_bytes_into(rem, memo)?;

        // Asset Identifier
        let asset_id = unsafe { &mut *addr_of_mut!((*out).asset_id).cast() };
        let rem = AssetIdentifier::from_bytes_into(rem, asset_id)?;
        let _asset_id = unsafe { asset_id.assume_init() };

        // PublicAddress
        let sender = unsafe { &mut *addr_of_mut!((*out).sender).cast() };
        let _rem = PublicAddress::from_bytes_into(rem, sender)?;

        unsafe {
            addr_of_mut!((*out).randomness).write(randomness);
            addr_of_mut!((*out).value).write(value);
        }

        Ok(())
    }

    ///// Verify that the note's commitment matches the one passed in
    //pub(crate) fn verify_commitment(&self, commitment: Scalar) -> Result<(), IronfishError> {
    //    if commitment == self.commitment_point() {
    //        Ok(())
    //    } else {
    //        Err(IronfishError::InvalidCommitment)
    //    }
    //}
    //
    ///// Compute the commitment of this note. This is essentially a hash of all
    ///// the note values, including randomness.
    /////
    ///// The owner can publish this value to commit to the fact that the note
    ///// exists, without revealing any of the values on the note until later.
    //pub(crate) fn commitment_point(&self) -> Scalar {
    //    // The commitment is in the prime order subgroup, so mapping the
    //    // commitment to the u-coordinate is an injective encoding.
    //    jubjub::ExtendedPoint::from(self.commitment_full_point())
    //        .to_affine()
    //        .get_u()
    //}
}
