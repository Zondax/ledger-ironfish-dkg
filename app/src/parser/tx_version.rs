use super::ParserError;

#[derive(Copy, PartialEq, Clone)]
pub enum TransactionVersion {
    V1,
    V2,
}

impl TransactionVersion {
    pub fn as_str(&self) -> &'static str {
        match self {
            TransactionVersion::V1 => "V1",
            TransactionVersion::V2 => "V2",
        }
    }
}

impl TryFrom<u8> for TransactionVersion {
    type Error = ParserError;
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            1 => Ok(TransactionVersion::V1),
            2 => Ok(TransactionVersion::V2),
            _ => Err(ParserError::InvalidTxVersion),
        }
    }
}
