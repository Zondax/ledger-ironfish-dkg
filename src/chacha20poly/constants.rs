pub const NONCE_LEN: usize = 12;
pub const KEY_LEN: usize = 32;
