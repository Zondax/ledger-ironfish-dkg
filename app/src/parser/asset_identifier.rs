use core::fmt::{self, Display, Formatter};
use core::ptr::addr_of_mut;

use arrayref::array_ref;
use nom::bytes::complete::take;

use crate::FromBytes;

use crate::parser::constants::ASSET_ID_LENGTH;

/// A convenience wrapper around an asset id byte-array, allowing us to push the
/// error checking of the asset id validity to instantiation
/// instead of when trying to get the generator point. This causes code relating
/// to notes and value commitments to be a bit cleaner
#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
pub struct AssetIdentifier([u8; ASSET_ID_LENGTH]);

impl AssetIdentifier {
    pub fn as_bytes(&self) -> &[u8; ASSET_ID_LENGTH] {
        &self.0
    }
}

impl<'a> FromBytes<'a> for AssetIdentifier {
    #[inline(never)]
    fn from_bytes_into(
        input: &'a [u8],
        out: &mut core::mem::MaybeUninit<Self>,
    ) -> Result<&'a [u8], nom::Err<crate::parser::ParserError>> {
        let (rem, raw) = take(ASSET_ID_LENGTH)(input)?;
        let bytes = array_ref!(raw, 0, ASSET_ID_LENGTH);

        let out = out.as_mut_ptr();

        unsafe {
            addr_of_mut!((*out).0).write(*bytes);
        }

        Ok(rem)
    }
}
