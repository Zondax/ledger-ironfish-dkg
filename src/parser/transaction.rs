use core::{mem::MaybeUninit, ptr::addr_of_mut};

use nom::{
    bytes::complete::take,
    number::complete::{le_i64, le_u32, le_u64, le_u8},
};

use crate::parser::constants::{KEY_LENGTH, REDJUBJUB_SIGNATURE_LEN};

mod burns;
mod mints;
mod outputs;
mod spends;

use super::{FromBytes, ObjectList, TransactionVersion};
pub use burns::Burn;
pub use mints::Mint;
pub use outputs::Output;
pub use spends::Spend;

// parser_error_t _read(parser_context_t *ctx, parser_tx_t *v) {
//     CHECK_ERROR(readTransactionVersion(ctx, &v->transactionVersion));
//     CHECK_ERROR(readUint64(ctx, &v->spends.elements));
//     CHECK_ERROR(readUint64(ctx, &v->outputs.elements));
//     CHECK_ERROR(readUint64(ctx, &v->mints.elements));
//     CHECK_ERROR(readUint64(ctx, &v->burns.elements));
//     CHECK_ERROR(readInt64(ctx, &v->fee));
//     CHECK_ERROR(readUint32(ctx, &v->expiration));
//
//     v->randomizedPublicKey.len = KEY_LENGTH;
//     CHECK_ERROR(readBytes(ctx, &v->randomizedPublicKey.ptr, v->randomizedPublicKey.len));
//
//     v->publicKeyRandomness.len = KEY_LENGTH;
//     CHECK_ERROR(readBytes(ctx, &v->publicKeyRandomness.ptr, v->publicKeyRandomness.len));
//
//     // Read Spends and Outputs
//     CHECK_ERROR(readSpends(ctx, &v->spends));
//     CHECK_ERROR(readOutputs(ctx, &v->outputs));
//
//     // Read Mints and Burns
//     CHECK_ERROR(readMints(ctx, &v->mints, v->transactionVersion));
//     CHECK_ERROR(readBurns(ctx, &v->burns));
//
//     v->bindingSignature.len = REDJUBJUB_SIGNATURE_LEN;
//     CHECK_ERROR(readBytes(ctx, &v->bindingSignature.ptr, v->bindingSignature.len));
//
//     if (ctx->bufferLen != ctx->offset) {
//         return parser_unexpected_buffer_end;
//     }
//
//     CHECK_ERROR(transaction_signature_hash(v, v->transactionHash));
//     return parser_ok;
// }

#[cfg_attr(test, derive(Debug))]
#[derive(Copy, PartialEq, Clone)]
pub struct Transaction<'a> {
    tx_version: TransactionVersion,
    random_pubkey: &'a [u8; KEY_LENGTH],
    pubkey_randomness: &'a [u8; KEY_LENGTH],

    spends: ObjectList<'a, Spend<'a>>,
    outputs: ObjectList<'a, Output<'a>>,
    mints: ObjectList<'a, Mint<'a>>,
    burns: ObjectList<'a, Burn<'a>>,
    fee: i64,
    expiration: u32,
    binding_sig: &'a [u8; REDJUBJUB_SIGNATURE_LEN],
}

impl<'a> FromBytes<'a> for Transaction<'a> {
    #[inline(never)]
    fn from_bytes_into(
        input: &'a [u8],
        out: &mut core::mem::MaybeUninit<Self>,
    ) -> Result<&'a [u8], nom::Err<super::ParserError>> {
        let out = out.as_mut_ptr();

        let (rem, raw_version) = le_u8(input)?;
        let version = TransactionVersion::try_from(raw_version)?;
        // now read the number of spends, outputs, mints and burns
        let (rem, num_spends) = le_u64(rem)?;
        let (rem, num_outputs) = le_u64(rem)?;
        let (rem, num_mints) = le_u64(rem)?;
        let (rem, num_burns) = le_u64(rem)?;
        // now read the fee and expiration
        let (rem, fee) = le_i64(rem)?;
        let (rem, expiration) = le_u32(rem)?;

        // This fields bellows are present in C parser, we need to figure out where to
        // place this information
        // rondomizedPublicKey
        let (rem, random_pubkey) = take(KEY_LENGTH)(rem)?;
        // publicKeyRandomness
        let (rem, randomness) = take(KEY_LENGTH)(rem)?;

        let random_pubkey = arrayref::array_ref![random_pubkey, 0, KEY_LENGTH];
        let pubkey_randomness = arrayref::array_ref![randomness, 0, KEY_LENGTH];

        let spends: &mut MaybeUninit<ObjectList<'a, Spend<'a>>> =
            unsafe { &mut *addr_of_mut!((*out).spends).cast() };
        let rem = ObjectList::new_into_with_len(rem, spends, num_spends as usize)?;

        let outputs: &mut MaybeUninit<ObjectList<'a, Output<'a>>> =
            unsafe { &mut *addr_of_mut!((*out).outputs).cast() };
        let rem = ObjectList::new_into_with_len(rem, outputs, num_outputs as usize)?;

        let mints: &mut MaybeUninit<ObjectList<'a, Mint<'a>>> =
            unsafe { &mut *addr_of_mut!((*out).mints).cast() };
        let rem = ObjectList::new_into_with_len(rem, mints, num_mints as usize)?;

        let burns: &mut MaybeUninit<ObjectList<'a, Burn<'a>>> =
            unsafe { &mut *addr_of_mut!((*out).burns).cast() };
        let rem = ObjectList::new_into_with_len(rem, burns, num_burns as usize)?;

        let (rem, sig) = take(REDJUBJUB_SIGNATURE_LEN)(rem)?;
        let binding_sig = arrayref::array_ref![sig, 0, REDJUBJUB_SIGNATURE_LEN];

        unsafe {
            addr_of_mut!((*out).tx_version).write(version);
            addr_of_mut!((*out).fee).write(fee);
            addr_of_mut!((*out).expiration).write(expiration);
            addr_of_mut!((*out).binding_sig).write(binding_sig);
            addr_of_mut!((*out).random_pubkey).write(random_pubkey);
            addr_of_mut!((*out).pubkey_randomness).write(pubkey_randomness);
        }

        Ok(input)
    }
}

impl<'a> Transaction<'a> {
    pub fn num_spends(&self) -> usize {
        self.spends.iter().count()
    }

    pub fn num_outputs(&self) -> usize {
        self.outputs.iter().count()
    }

    pub fn num_mints(&self) -> usize {
        self.mints.iter().count()
    }

    pub fn num_burns(&self) -> usize {
        self.burns.iter().count()
    }
}

#[cfg(test)]
mod transaction_test {
    use crate::*;

    // testing data for unit tests
    const TEST_TX: &str = "02010000000000000004000000000000000100000000000000010000000000000001000000000000000000000077b55de659bda8a
31c3627b9270139637c1fe55efc1cfc09dbb697d68d7583b6844339d824d73947aa6bdafac81c65910da301837980a57cbb1243c223192446494e0770fe825d36
eb783ecc0ecc8c80b1ebb8915ae9517d4070ba5613bc23d4a975beeab6dc63e153830eda6a755b4986fbdd4449724b87218045de587f5b6c0991c3a8f88ae5a67
a77a2f36fd63c58bc770a86760175b2af5213cd3a7bfebfa6b52508acc97a3f1b78e4be166aa6868f19995318ef578855546250957becf6ac79c1abb60962854e
61ef3c4d35eccc0ac10c06bb31881bf9335cca1318509b7cd1f38ed7168e444569ecb97a5fcfa416f61f5aada0cb73d185723dfe1ad7508e5a74e170838290db9
657ccbed94b2828d43b31b5abf89af0ed3a3c0caee460780500002f250c30c12c53a475af23d3d6419a6522e350f84d603e82f1b5aa661423c2d878cb1f72f28a
40854402380066a711300db6ccc70182a1231e57fee03f32e131630dddc58592b032a4e48cea93798c894b8d35b16495b4298326aff034d41700841ca42e0e443
e29bee9bd6b072b2181376acd607093fd18daaf33087c354aa3d02d56c8c4eb9a92e40db161d44d92e9b65efd9ff9fac0a42399e2d459e9d4e29abfb1c3fa0e9b
ab8d6aeb8177d2c73393f5501ff56171e7ec94f2ef7c8d1b4c18faf62e50930eb25d42bcea4e4a01c75d610e2770fc3e3a3eb1e7779bc7f4306fd485eae3547c1
2223cc1292ee6c6dd9610ef4fcebf08693a6a76363419e70326e1ac088466d4ed51e53c997abd606840492a7486730d72eb4ba6e4013006e91237100f692657ea
6f3a60de2a061e40b1798fdd49e6edb2f03f9719804dab4ea54080718898b0ef77fed41cdf0417de59db823d43fe8fbcb3360da6c24267264f28f8e485f81839a
0eef52322f67e56ea6a0d0f0f77e3a7f3de73ab92a67a82f7d00ff619aecb73c6e9960c8cf876ca3b242fcec567ab5066fea40e666eed3a9336ba9599b17b5483
d680fe57f51dabfecda43fe4e7ed600e549191697e3b9fef0937bedb680fe7daa93ba4b8523e0cbf2d442aa0105c97d66591c2cfe61b71396c1410ae819fe7c5b
6754ff9d855576c670b93242ab5f8188e01f31e567036eedc0aca8edb4e3f6fbc0606207a9f452199bfbb60dcf3c2e648ee47238c8ac602a23a316a20e51c4a04
d0a1770cf0e11c4773e7c83b1b904ddd4c07a250826730e29ca9bc335b46c58910f767675b6988a5edd8c0abd0bbb744003f0fdc1ea703d50ff141380c7c838cc
b12a2faf1041eac257261a306491ff098e3c29fbf8ea5ac909497cb00dafbcb1be53fe494a76926ee9428f8c5bd846e24c9e08aaa6aaa0de4c041c83adac515c0
661a8d757226fbc03270e9633ec1aa312be7165ede814bcc3ef9a468ac112aa175841118c8f20b8d268b5b8cbdee72913cc043c81f2c5058dbae899669428de41
3c1b811e03f327e6caaff02728af22c72227950cf5653822c160cbc2adc5075a4eaa2d46d201e01b8ef8a69d9ae898a395e9800143f68fe0627abbf70c65cfc22
8c403408b3931461f088e925dbd5b4d5ccb0ec98bd8847aa1801628b76a24734ce7dc77b2f470969c95fc0d110cc473a640ea3146d57b4d4d3931d3bdbd3851c5
0cf0982538b5f84eb889e89b2def189767d1e4d3b416b470636b10402183c7c711a2219783dc7c63dbac78335e42383d3f2495ae5eed81cb6ddc345a9c24449cb
3d6ab468f15ce2a24a829b568b5890172ec2ef58f3e895143926ccecb9d85141ee819207173dc2acebd97a89a23c0f9f63ae490f3dd52efb4c2b58cc4be254df4
8d8e5724e1f5e640b95c745508734be2f50ecb2100a09c1f7bb65accc6d54b1e2ef801bd39f677147566d61cd81041b76ff589dadb04a696ac3f1e2f904593231
abe4b0b1d933e80b20bcf65a54fdc1b39a986a2470cd17d003f5f6f52773b5e99cd55f823783758b46ef86e48caa3b2177e69abc4217be50444efc52dc9407a65
29aa514c577363b52eafed2c78e7fd81869f80b36577d513b319546756a80d9e829c4ccfe5f1abbcdb330d935a6e4ceb7eb98d8c8425ce7000743ba0b41793f82
00349afe93c0799fc4bfe3e3d381df69a90c700a51cac12b2456f56593457e7ffb0ec63680ec1846a7bc43761a566864d2f0854d419727d415b99c2d1a7abc460
f167541f30e822f977e022134ab8c24f9acb8a721da467e661fc745766553063ca6b880a19fcf929063ada26c6f38da00d0757a917c64759a90d92f3d4cb0fd78
b55197780aec0415695d78b55669953fd258b5f05b1dcb6eecc47b410cf7ed06c4e57cb8f34f180d36e218c24acfc7d73c8c8e5ffb2bc2c938bfaba539f3972a0
b9f515c8e7fbae327087364b4acaedde75c920cbf543efb17671204bd9f72437478cddad295ca5b70c44f8ed34d4d211ec18b65fb08a4ae5181e0a07deb45d0d2
7d2b864dd5557cf179d824671bc22d4e2017e4a9858b2fc62a525bc0eb2aa4a2f9d0196f13935502b977c4f56fc61f3932cde5f63b88206a59876923ec4a5169b
e61e75d4f7ecb1e1dd13e0c3798329aa890dddc6391e4a42948420b205c4752a0bd5feb016f1b78bf8dd5c5cae0696fc5697f4b47a4c0463e71e0d401612a2df5
b5ca68618a0032de25847b6f305d0051cb092cde82f17cf806088de4c8ff95d8a00b6fec4e88436d1608718905c2fdef39e87d7bf789dc079570625516bca69bd
7b6adb1dd1d9300c933fa4c74c33cd06178b095704cf4d67ceda6a397bf753918e0205f6041f714b4d2ed6a7063111cbe434ffa7018b71793352626346bf8e268
04716f261ef6a4b11d73359e19eb566b2f97d1831a12b30e091b779442a219f8444137ddda1691487939f61fb7819e9f885193ee54bc499edbb42b9abe48944fb
363e2aa01a26cf090b79ec2f113038d745656ed2f98af671640098dfae56769748ba96d6aed11c7e8d4e91b1f2bcbf8dfe6f3cb93caffd4c7bc24a3b8dddd6c75
5652b9a7c632876fb2eb9e28405295147c7fc39ad059f71b15a1d164594e272b36bede5ef080c4d3de38388fb8253067c1b5877290a66bde5cc994576d61fd319
93337224531e2626dac76104c70ea3bd10855756a16362466fdf8cacc64d4d96e9f5c5b07f2dfd53567a56d45d97bc9abd507705f6e3bbad3a99aa3e921ce8331
4e60b9ae34221348223d115e45fa1bea7ead2694c4898792b1e6467efaad90df2e6806eb454658e8236cc773bcc9fde10127b67cd5917dfd17a9cce7eda1c1adf
d0ebd43f93146f6d064dfa58a6e6b83737e8cffc4c717215edb22c5ca812adf1f52e9ed803bc2f0a303cbda0bf5175632b6c0fb384f873fcbb197d13440e4bdad
9a41475e4514234bcc4e2bfc482c153771a5eb22950b50e6370dc838b28457cd3eb41fe31a340270c6018d486580b69bd82f164e62dfd0e1b81653006b89861e9
1221f9ca5af0e2a35df8ea65c90c5ce20e299beae7cec8ec88c6e3b1bd30f8eec812166997102fa4ecff0c3e6287c66b5e4a629eceef9f7d225b4175697bd7421
acb8dc05e7891e3dc94dc13841feb3b0ad804a11d24896ce4cb07a46bf6888e4217c52e94c969cd58a0f64d134257326e81b57d9c1ffd0ade07f550a52ac96f2b
3d51eeb47e5325cbcfbd27df221b4286ffe49fd7491bcea88a0f70d3e312f4ccf7183928232fa29061e9b881fbc34c15d9f1a39ad069efde9e7fa892345d0dafa
21fca8408d720d6d79e0167e8ab13695f6d2c787545dad56f091cfb786e0f4459213497577e9cb9a54657374636f696e000000000000000000000000000000000
00000000000000041207265616c6c7920636f6f6c20636f696e000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000004050000000000000079e0167e8ab13695f6d2c787545dad56f
091cfb786e0f4459213497577e9cb9a00b00ad4cdfd5c66e1822c22acd26da08c21cb366632a03210c364d33d259deee66c50da30ba43bbcedcf1aff4370a0191
2069aca1355d360e5806653b8656820c6d9e3dd19d175b3eab6cb439d69f19b25a95b9f2c3270c23543578b3667c8ea20200000000000000280d529a652296987
bce8140f0d72d548ccec277dd4add9c5f378d56178f86cfa230822819e6ae9f1bdb2ec9c00dbc954dc52e0730157860dd41a666213fea03";

    #[test]
    fn parse_tx() {
        let tx = hex::decode(TEST_TX).unwrap();
        let (_, tx) = Transaction::from_bytes(&tx).unwrap();
        assert_eq!(tx.num_spends(), 1usize);
        assert_eq!(tx.num_outputs(), 1usize);
        assert_eq!(tx.num_mints(), 1usize);
        assert_eq!(tx.num_burns(), 1usize);
    }
}
