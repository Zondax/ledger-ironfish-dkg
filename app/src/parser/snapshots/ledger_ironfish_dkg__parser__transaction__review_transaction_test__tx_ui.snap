---
source: src/parser/transaction.rs
expression: reduced
input_file: src/parser/testvectors/transaction_simple.json
---
[
    "Tx Version": "V1",
    "Owner 1": "40fae059d8ee3361b7a08429867254523f937c7654ae6fe7b188c1f0da57e9cd",
    "Amount 1": "40",
    "AssetID 1": "51f33a2f14f92735e562dc658a5639279ddca3d5079a6d1242b2a588a9cbf44c",
    "Owner 2": "40fae059d8ee3361b7a08429867254523f937c7654ae6fe7b188c1f0da57e9cd",
    "Amount 2": "5",
    "AssetID 2": "da40fc530a12327a1c1c367b8bd66ada35100501228cdb355b8fb6c40d048a64",
    "Owner 3": "40fae059d8ee3361b7a08429867254523f937c7654ae6fe7b188c1f0da57e9cd",
    "Amount 3": "1",
    "AssetID 3": "51f33a2f14f92735e562dc658a5639279ddca3d5079a6d1242b2a588a9cbf44c",
    "Fee": "1",
    "Expiration": "0",
]
